SEBrowser/Controllers/OpenXDAController.cs

```diff
@@ -119,178 +119,253 @@ public DataTable GetEventSearchData(EventSearchPostData postData)
-                    else
-                        eventCharacteristicsRestricitons += $" AND ((SELECT d.PerUnitMagnitude FROM Disturbance d WHERE d.ID = EventWorstDisturbance.WorstLLDisturbanceID) > {postData.transientMin} OR (SELECT d.PerUnitMagnitude FROM Disturbance d WHERE d.ID = EventWorstDisturbance.WorstLNDisturbanceID) > {postData.transientMin} OR (SELECT d.EventTypeID FROM Disturbance d WHERE d.ID = EventWorstDisturbance.WorstDisturbanceID) IN (SELECT ID FROM EventType WHERE Name <> 'Transient'))";
-                }
+                string query = $@" 
+	                SELECT 
+                        [SEBrowser.EventSearchEventView].*,
+                        [SEBrowser.EventSearchDetailsView].*
+                    FROM (SELECT TOP 100 * FROM ( SELECT
+	                    Event.ID as EventID,
+                        EventWorstDisturbance.WorstDisturbanceID AS DisturbanceID,
+                        NULL AS FaultID
+	                FROM
+	                    Event INNER JOIN
+	                    EventWorstDisturbance ON EventWorstDisturbance.EventID = Event.ID  Left JOIN
+                        FaultSummary ON FaultSummary.ID IS NULL
+                    WHERE {getTimeFilter(postData)} {filters}                       
+                    UNION
+                    SELECT Event.ID as EventID,
+                        NULL AS DisturbanceID,
+                        NULL AS FaultID
+	                FROM
+	                    Event LEFT JOIN
+                        FaultSummary ON FaultSummary.ID IS NULL LEFT JOIN
+                        EventWorstDisturbance ON EventWorstDisturbance.ID IS NULL
+                    WHERE {getTimeFilter(postData)} AND EventTypeID IN (SELECT ID FROM EventType WHERE NAME IN ('BreakerOpen','Other')) {filters}
+                    UNION
+                    SELECT Event.ID as EventID,
+                        NULL AS DisturbanceID,
+                        FaultSummary.FaultNumber AS FaultID
+	                FROM
+	                    Event INNER JOIN 
+                        FaultSummary  ON Event.ID = FaultSummary.EventID AND FaultSummary.IsSelectedAlgorithm <> 0 AND FaultSummary.IsValid <> 0 AND FaultSummary.IsSuppressed = 0 LEFT JOIN
+                        EventWorstDisturbance ON EventWorstDisturbance.ID IS NULL
+                    WHERE {getTimeFilter(postData)} {filters}
+                    ) X) Main LEFT JOIN
+                    [SEBrowser.EventSearchEventView] ON Main.EventID =  [SEBrowser.EventSearchEventView].EventID Inner JOIN    
+                    [SEBrowser.EventSearchDetailsView] ON
+                        [SEBrowser.EventSearchDetailsView].DisturbanceID = Main.DisturbanceID AND (
+                        [SEBrowser.EventSearchDetailsView].EventID = Main.EventID OR 
+                        [SEBrowser.EventSearchDetailsView].FaultID = Main.FaultID)
```

Comment:
This set of conditions feels wrong to me, but I'm not 100% sure what you were going for. Maybe something more like this?

```suggestion
                        (Main.DisturbanceID IS NOT NULL AND [SEBrowser.EventSearchDetailsView].DisturbanceID = Main.DisturbanceID) OR
                        (Main.FaultID IS NOT NULL AND [SEBrowser.EventSearchDetailsView].EventID = Main.EventID AND [SEBrowser.EventSearchDetailsView].FaultID = Main.FaultID) OR
                        (COALESCE([SEBrowser.EventSearchDetailsView].DisturbanceID, Main.DisturbanceID) IS NULL AND COALESCE([SEBrowser.EventSearchDetailsView].FaultID, Main.FaultID) IS NULL AND [SEBrowser.EventSearchDetailsView].EventID = Main.EventID)
```

Reply:
@clackner-gpa Any chance you can give this a quick look before tomorrow?

Reply:
The reason this is so complex is because you can have a Disturbance and a Fault on the same Event. The original logic says JOIN Event  using the `DisturbanceID` which covers all Disturbances independently of what else is going on in the Event. But then we ALSO need to join any additional Faults or Other Events since there could be some on events that are already joined to a Disturbance.  There is no way something that falls under OTHER is also a Fault (since both of these are determoned by the single value in `EventTypeID` so the `OR` works there). I think your logic does not account for the pretty common case of a FAULT having a Disturbance which would cause some issues

Reply:
If I'm not mistaken, the two result sets we are joining here are each the result of the `UNION` of three mutually exclusive cases: one that joins event and disturbance, one that joins event and fault, and the last that contains only the event ID for the remaining event types. Since they are mutually exclusive sets, no single record should include all three IDs. Is that wrong?

Reply:
ok the Inner Query (Main) returns a set of 3 IDS where everything has an `EventID` . There are some cases with a `FaultID`, some cases with a `DisturbanceID` and some with `NULL` in both. Take the following 3 Events:
| Event ID | Event.EventType |
|----------|-----------------|
| Event 1  | Fault           |
| Event 2  | Sag             |
| Event 3  | Other           |

And the following Disturbances:
| Disturbance ID | Disturbance.EventType | Associated Event |
|----------|-----------------|-----------------|
| Dist 1  | Transient | 1 |
| Dist 2  | Sag          | 2 |

The only Fault would be `Fault 1` which is associated with `Event 1`

The inner Select (`Main`) will return
| Event ID | Disturbance ID | Fault ID |
|----------|----------------|----------|
| 1        | 1              | NULL     |
| 2        | 2              | NULL     |
| 3        | NULL           | NULL     |
| 2        | NULL           | 1        |

Where Rows 1 and 2 are from the disturbance part `UNION` to get row 3 from the `OTHER` part and `UNION` row 4 from the Fault part.

now the current Logic (as is) will do a `JOIN` based on 2 Conditions with an AND in between. 
(1) `DisturbanceID` matching
(2) either `EventID ` or `FaultID` matching.

Looking at this example, the `EventSearchDetailsView` in openXDA would look something like this:

| RowID  | Event ID | Disturbance ID | Fault ID | Details for Fault | Details for Disturbance | Details for Event |
|--------|----------|----------------|----------|-------------------|-------------------------|-------------------|
| 1      | 1        | 1              | NULL     | N/A               | Disturbance 1           | N/A               |
| 2      | 2        | 2              | NULL     | N/A               | Disturbance 2           | N/A               |
| 3      | 2        | NULL           | 1        | Fault 2           | N/A                     | N/A               |
| 4      | 3        | NULL           | NULL     | N/A               | N/A                     | Event 2           |

Walking through the current `JOIN` statement Row by row we get:
`Main` row 1 is matched with `EventSearchDetailsView` row 1 since that is the only one that satisfies condition (1) `AND` (2) based on the `DisturbanceID` and `EventID`

`Main` row 2 is matched with `EventSearchDetailsView` row 2 . Condition (2) is matched by row 2 and row 3, but condition (1) prevents row 3 from being joined.

`Main` row 3 is matched with `EventSearchDetailsView` row 3 . Condition (1) is matched by row 3 and row 4, but condition (2) only matches row 3.

`Main` row 4 is matched with `EventSearchDetailsView` row 4 . Condition (1) is matched by row 3 and row 4, but condition (2) only matches row 4.

Looking at that logic this seems to work similarly with your suggestion. The two cases that come to mind otherwise are 2 Event's with the same `FaulID` since we are using the `FaultNumber` for that this may be an issue. I'll work through that in a minute.

Reply:
Walking though your join we have 3 conditions that are combined via `OR` so only one needs to be met:
(1) `DisturbanceID` not `NULL` and it matches
(2)  `FaultID ` not `NULL` and `EventID` and `FaultID` both match
(3) `DisturbanceID` is `NULL` in either row and `FaultID` is null on either and `EventID matches

Doing the joining: 
`Main` row 1 is matched with `EventSearchDetailsView` row 1 due to condition (1), none of the other conditions are true.

`Main` row 2 is matched with `EventSearchDetailsView` row 2 due to condition (1), none of the other conditions are true.

`Main` row 3 is matched with `EventSearchDetailsView` row 4 due to condition (3), none of the other conditions are true.

`Main` row 4 is matched with `EventSearchDetailsView` row 4 due to condition (2), none of the other conditions are true.

so that seems to work ok either way

Reply:
Now looking at the case with 2 Events and the same `FaultID` (and for good measure, what if it's a `BreakerOpen` with a `transient` since that's also a case I just found:

| Event ID | Event Type | 
|--------|----------|
| Event 1      | Fault       | 
| Event 2      | Fault        |
| Event 3      | Breaker Open |

| Fault ID | Descr (unique) | Associated Event |
|--------|----------|------|
| Fault 1      | Fault A |  1
| Fault 1      | Fault  B      | 2 |

There is a single Disturbance for Event 3.

The inner Join will result in
| Event ID | Disturbance ID | Fault ID |
|--------|----------|------|
| 3     | 3 |  NULL |
| 3     | NULL |  NULL |
| 1     | NULL | 1 |
| 2    | NULL  | 1 |

The `EventSearchDetailsView` looks something like:
| RowID  | Event ID | Disturbance ID | Fault ID | Details for Fault | Details for Disturbance | Details for Event |
|--------|----------|----------------|----------|-------------------|-------------------------|-------------------|
| 1      | 1        | NULL              | 1 |  Fault A              | N/A           | N/A               |
| 2      | 2        | NULL              | 1|  Fault B            | N/A           | N/A               |
| 3      | 3        | 1 | NULL        | N/A          | Disturbance 1                     | N/A               |
| 4      | 3        | NULL           | NULL     | N/A               | N/A                     | Event 3           |

Joins results are: 
`Main` row 1 joins with row 3 due to (1), no other rows match
`Main` row 2 joins with row 4 due to (3), no other rows match
`Main` row 3 joins with row 1 due to (2), no other rows match since row 2 has a different `eventID`
`Main` row 4 joins with row 2 due to (2), no other rows match since row 1 has a different `eventID`

The old logic breaks down in this case so that's out the window

Reply:
Is there a possibility of the `EventType` being `Fault` or `RecloseIntoFault` but no `FaultSummary` available?

Reply:
> Looking at this example, the `EventSearchDetailsView` in openXDA would look something like this:
> ...

FYI, I believe this is wrong because `NULL = NULL` evaluates to false. The old logic would return only the first two rows in your example. More generally, it returns only the rows where `DisturbanceID IS NOT NULL` because `[SEBrowser.EventSearchDetailsView].DisturbanceID = Main.DisturbanceID` is a necessary condition.

Strictly speaking, the new logic doesn't need to check for `NULL` in the first two cases. However, I think it makes the logic a bit clearer since the conditions are explicitly crafted to appear mutually exclusive.

> Is there a possibility of the `EventType` being `Fault` or `RecloseIntoFault` but no `FaultSummary` available?

Only if the records are manually updated. There are several places throughout the codebase where we assume that every fault must have a fault summary so I wouldn't worry too much about it.

---

Thanks for the in-depth analysis here. If you're happy with it, then we should be good to go.

Reply:
I am almost positive there was an issue I tried to resolve I just can't remember what that was. I hope the extensive comments will help in resolving it in the future - if we ever figure out what it was.